src/pool/srv_pool.c

```diff
@@ -1274,21 +1279,64 @@ read_db_for_stepping_up(struct pool_svc *svc, struct pool_buf **map_buf,
 				     RAS_SEV_ERROR, NULL /* hwid */,
 				     NULL /* rank */, NULL /* inc */,
 				     NULL /* jobid */,
-				     &svc->ps_uuid, NULL /* cont */,
+				     &uuid_tmp, NULL /* cont */,
 				     NULL /* objid */, NULL /* ctlop */,
 				     NULL /* data */,
 				     "incompatible layout version: %u larger than "
 				     "%u", global_version,
 				     DS_POOL_GLOBAL_VERSION);
 		rc = -DER_DF_INCOMPT;
-		goto out_lock;
+		goto out_map_buf;
 	}
 
-	rc = pool_prop_read(&tx, svc, DAOS_PO_QUERY_PROP_ALL, prop);
+	D_ASSERTF(rc == 0, DF_RC"\n", DP_RC(rc));
+	*map_buf_out = map_buf;
+	*map_version_out = map_version;
+out_map_buf:
 	if (rc != 0)
-		D_ERROR(DF_UUID": cannot get access data for pool: "DF_RC"\n",
+		D_FREE(map_buf);
+out:
+	return rc;
+}
+
+/*
+ * Read the DB for map_buf, map_version, and prop. If the return value is 0,
+ * the caller is responsible for freeing *map_buf_out and *prop_out eventually.
+ */
+static int
+read_db_for_stepping_up(struct pool_svc *svc, struct pool_buf **map_buf_out,
+			uint32_t *map_version_out, daos_prop_t **prop_out)
+{
+	struct rdb_tx		tx;
+	struct pool_buf	       *map_buf;
+	uint32_t		map_version;
+	daos_prop_t	       *prop;
+	int			rc;
+
+	rc = rdb_tx_begin(svc->ps_rsvc.s_db, svc->ps_rsvc.s_term, &tx);
+	if (rc != 0)
+		goto out;
+	ABT_rwlock_rdlock(svc->ps_lock);
+
+	rc = ds_pool_svc_load(&tx, svc->ps_uuid, &svc->ps_root, &map_buf, &map_version);
+	if (rc != 0)
+		goto out_lock;
+
+	rc = pool_prop_read(&tx, svc, DAOS_PO_QUERY_PROP_ALL, &prop);
+	if (rc != 0) {
+		D_ERROR(DF_UUID": failed to read pool properties: "DF_RC"\n",
 			DP_UUID(svc->ps_uuid), DP_RC(rc));
+		daos_prop_free(prop);
```

Comment:
does prop need to be initialized to NULL in this function (or is it assumed that the compiler will do this in the build)? If the pool_prop_read() code fails in the rdb_tx_lookup() of ds_pool_prop_global_version, the prop will not have been allocated yet.

Reply:
Indeed. This is a bug.